### ISSUE TITLE
Upgrade to stable-2.10.0 from stable-2.9.4 causes “Could not fetch profile” errors

### ISSUE THREAD

**Issue by cpackingham** [linkerd/linkerd2#5881]
### What is the issue?
There is a significant volume of "Could not fetch profile" WARN logs after upgrading. Over 2 million were produced in under an hour. 

### How can it be reproduced?
Upgrade to stable-2.10.0 from stable-2.9.4 on EKS. Inject a pod which is sending traces to the datadog agent, while leaving the agent uninjected. Make port 8126 (the agent trace port) opaque.

### Logs, error output, etc
![image (6)](https://user-images.githubusercontent.com/31993057/110503426-c8bff980-80c1-11eb-9e8b-428a51e37a34.png)
![Screen Shot 2021-03-08 at 3 39 13 PM](https://user-images.githubusercontent.com/31993057/110503478-d8d7d900-80c1-11eb-93e7-ab636fcc931e.png)
![patterns](https://user-images.githubusercontent.com/31993057/110503744-1fc5ce80-80c2-11eb-9a5a-aebaa3375cf5.png)

#### `linkerd check` output

```text
~/Downloads
(⎈ |dev:dev)❯❯❯ ./linkerd2-cli-edge-21.3.1-darwin check
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API
kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version
linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist
linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2021-04-24T03:34:47Z
    see https://linkerd.io/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints
√ issuer cert is issued by the trust anchor
linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days
linkerd-api
-----------
√ control plane pods are ready
√ can initialize the client
√ can query the control plane API
linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date
control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match
Status check results are √
Linkerd extensions checks
=========================
linkerd-viz
-----------
√ linkerd-viz Namespace exists
√ linkerd-viz ClusterRoles exist
√ linkerd-viz ClusterRoleBindings exist
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ tap API service is running
√ linkerd-viz pods are injected
√ viz extension pods are running
√ prometheus is installed and configured correctly
√ can initialize the client
√ viz extension self-check
Status check results are √
```

### Environment

- Kubernetes Version: 1.16.15
- Cluster Environment: EKS
- Host OS: Debian
- Linkerd version: edge-21.3.1

**Comment by kleimkuhler** [comment]
@cpackingham Thanks for the report. There are a few questions I have while I try to reproduce this.

1. Did the upgrade occur when the warnings started? If not, do you know how long `edge-21.3.1` was installed before the warnings started?
2. Could you attach logs for the linkerd pods? I'm interested in logs for the destination pod as well as the logs for the pods where a lot of these warnings/errors are happening. Additionally with the logs, a `kubectl get -n linkerd pods -o wide` would be helpful to be able to lineup expected IPs.
3. Are these warnings coming from an application pod or a linkerd pod? If they are coming from an application pod, have those pods been restarted so that they get new proxy containers?
4. I'm trying to reproduce this but have not been able to. Is there anything unique about your installation? Do you see this on a demo app like [emojivoto](https://linkerd.io/2/getting-started/#step-5-install-the-demo-app) or [booksapp](https://linkerd.io/2/tasks/books/#install-the-app)?

**Comment by cpackingham** [comment]
@kleimkuhler 
1. The warnings started pretty immediately after the upgrade
2.
Results of `kubectl get -n linkerd pods -o wide`:
```
NAME                                      READY   STATUS    RESTARTS   AGE     IP               NODE                            NOMINATED NODE   READINESS GATES
linkerd-controller-59c9688d68-5tbmz       2/2     Running   0          7m25s   10.100.107.176   ip-10-100-106-7.ec2.internal    <none>           <none>
linkerd-destination-74c786dbf6-59spd      2/2     Running   0          7m25s   10.100.106.51    ip-10-100-106-7.ec2.internal    <none>           <none>
linkerd-identity-5d8c79c54-gwn26          2/2     Running   0          7m24s   10.100.103.25    ip-10-100-103-63.ec2.internal   <none>           <none>
linkerd-proxy-injector-69fcfdb468-5p9dr   2/2     Running   0          7m24s   10.100.100.192   ip-10-100-103-63.ec2.internal   <none>           <none>
linkerd-sp-validator-58bf7499df-wzfgb     2/2     Running   0          7m24s   10.100.102.223   ip-10-100-103-63.ec2.internal   <none>           <none>
```
Logs from destination pod proxy:
```
(⎈ |dev:dev)❯❯❯ k logs -n linkerd deploy/linkerd-destination linkerd-proxy
time="2021-03-09T21:52:33Z" level=info msg="running version edge-21.3.1"
{"timestamp":"[     1.043697s]","level":"WARN","fields":{"message":"Failed to proxy request: error trying to connect: Connection refused (os error 111)"},"target":"linkerd_app_core::errors","threadId":"ThreadId(1)"}
{"timestamp":"[    11.700962s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    15.813764s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    42.673240s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    43.417407s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    45.796843s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    47.417519s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    47.574779s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[    49.823331s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[   162.003590s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[   165.483623s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[   362.216803s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
{"timestamp":"[   482.819334s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
```

Logs from destination pod:
```
(⎈ |dev:dev)❯❯❯ k logs -n linkerd deploy/linkerd-destination destination
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer ops/velero-5bd74c646f: replicaset.apps \"velero-5bd74c646f\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer ops/velero-5bd74c646f: replicaset.apps \"velero-5bd74c646f\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-sp-validator-58bf7499df: replicaset.apps \"linkerd-sp-validator-58bf7499df\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-sp-validator-58bf7499df: replicaset.apps \"linkerd-sp-validator-58bf7499df\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer ephemeral/binding-tc-bot-110-57c89c6765: replicaset.apps \"binding-tc-bot-110-57c89c6765\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer ephemeral/binding-tc-bot-110-57c89c6765: replicaset.apps \"binding-tc-bot-110-57c89c6765\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-controller-5bd44bddd6: replicaset.apps \"linkerd-controller-5bd44bddd6\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-controller-5bd44bddd6: replicaset.apps \"linkerd-controller-5bd44bddd6\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer dev/account-client-dev-6cf8579597: replicaset.apps \"account-client-dev-6cf8579597\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer dev/account-client-dev-6cf8579597: replicaset.apps \"account-client-dev-6cf8579597\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer dev/pixelon-dev-7d9f6d9f6b: replicaset.apps \"pixelon-dev-7d9f6d9f6b\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-prometheus-86b899b855: replicaset.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-prometheus-86b899b855: replicasets.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-prometheus-86b899b855: replicaset.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-prometheus-86b899b855: replicasets.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-web-55685cf58f: replicaset.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-web-55685cf58f: replicasets.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:35Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-web-55685cf58f: replicaset.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-web-55685cf58f: replicasets.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-grafana-6c786b7d6b: replicaset.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-grafana-6c786b7d6b: replicasets.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-grafana-6c786b7d6b: replicaset.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-grafana-6c786b7d6b: replicasets.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-tap-b5944d5f: replicaset.apps \"linkerd-tap-b5944d5f\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-tap-b5944d5f: replicasets.apps \"linkerd-tap-b5944d5f\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-web-55685cf58f: replicaset.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-web-55685cf58f: replicasets.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:36Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-prometheus-86b899b855: replicaset.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:37Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-prometheus-86b899b855: replicasets.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:37Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-grafana-6c786b7d6b: replicaset.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:37Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-grafana-6c786b7d6b: replicasets.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:37Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-prometheus-86b899b855: replicaset.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:37Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-prometheus-86b899b855: replicasets.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:38Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-grafana-6c786b7d6b: replicaset.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:38Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-grafana-6c786b7d6b: replicasets.apps \"linkerd-grafana-6c786b7d6b\" not found"
time="2021-03-09T21:52:42Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-web-55685cf58f: replicaset.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:42Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-web-55685cf58f: replicasets.apps \"linkerd-web-55685cf58f\" not found"
time="2021-03-09T21:52:46Z" level=warning msg="failed to retrieve replicaset from indexer linkerd/linkerd-prometheus-86b899b855: replicaset.apps \"linkerd-prometheus-86b899b855\" not found"
time="2021-03-09T21:52:46Z" level=warning msg="failed to retrieve replicaset from direct API call linkerd/linkerd-prometheus-86b899b855: replicasets.apps \"linkerd-prometheus-86b899b855\" not found"
```
I think these are just errors from when I upgraded to the edge version and the `linkerd-viz` chart wasn't installed for a bit.

3. They are coming from injected proxies alongside application pods. Here are some of the errors from a proxy:
```
{"timestamp":"[   125.266342s]","level":"WARN","fields":{"message":"Could not fetch profile","error":"status: Unknown, message: \"http2 error: protocol error: unexpected internal error encountered\", details: [], metadata: MetadataMap { headers: {} }"},"target":"linkerd_service_profiles::client","threadId":"ThreadId(1)"}
{"timestamp":"[   125.266364s]","level":"WARN","fields":{"message":"Could not fetch profile","error":"status: Unknown, message: \"http2 error: protocol error: unexpected internal error encountered\", details: [], metadata: MetadataMap { headers: {} }"},"target":"linkerd_service_profiles::client","threadId":"ThreadId(1)"}
```
Proxies on 2.9.4 do not have this stream of errors, only when pods are restarted and get the new proxy do these start happening.
4. I did a clean install of emojivoto on our cluster and injected it with the edge proxies and do not see the error. Are there any details I could get which might help, like a yaml of one of our injected pods?

**Comment by olix0r** [comment]
Are you able to run the `linkerd-destination` controller with debug logging?

I'd `kubectl edit -n linkerd deploy/linkerd-destination` and modify the proxy sidecar container to have

```yaml
      - env:
        - name: LINKERD2_PROXY_LOG
          value: linkerd=debug,warn
```

I'm primarily concerned about the errors we see in the destination controller's proxy. All of the other errors will be easier to understand when we understand why we're seeing errors like:
```
{"timestamp":"[     1.043697s]","level":"WARN","fields":{"message":"Failed to proxy request: error trying to connect: Connection refused (os error 111)"},"target":"linkerd_app_core::errors","threadId":"ThreadId(1)"}
{"timestamp":"[    11.700962s]","level":"WARN","fields":{"message":"Failed to resolve control-plane component","error":"end of stream reached"},"target":"linkerd_app_core::control","threadId":"ThreadId(1)"}
```

**Comment by gokcesurenkok** [comment]
same issue happened to me as well ,
```
[  2220.261532s]  WARN ThreadId(01) outbound:accept{client.addr=10.233.75.25:34556 target.addr=192.168.55.208:30621}: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[  2220.268280s]  WARN ThreadId(01) outbound:accept{client.addr=10.233.75.25:55724 target.addr=192.168.55.208:30660}: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
```

and i changed into debug logging in destination services as you said @olix0r 
i am attaching the logs
https://gist.github.com/gokcesurenkok/d4ce32f2ceb6f1b561032558ef45cee3

**Comment by cpackingham** [comment]
This might be a similar issue: https://github.com/linkerd/linkerd2/issues/4870#issuecomment-772869135

@kleimkuhler has been helping me debug over slack - we found out that this was caused because an injected pod was trying to send traces to uninjected datadog agent pods on port 8126. The solution for now is to continue ignoring that port (instead of migrating it to `opaque`) in 2.10.

**Comment by stathisbm** [comment]
I noticed the same issue with kube-prometheus-stack in the linkerd-proxy on `prometheus-kube-prometheus-stack-prometheus-0` 
First the error encountered because the node-exporter daemonsets has the hostNetwork value to true. I changed it to false so the injection can happen (not sure if that was a good practice or not). Then the logs got the same error:

```[   648.067099s]  WARN ThreadId(01) outbound:accept{client.addr=192.168.241.91:37228 target.addr=192.168.200.5:10249}:poll_profile: linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }```

The only difference is that now the target_addr is the ec2 node that hosts the pod. Still I am trying to fix it with no luck.

Anyone else encountered the same issue?
